decidim-initiatives/app/services/decidim/initiatives/signature_handler.rb

```diff
@@ -0,0 +1,222 @@
+        return if authorization_handler_form_class.blank?
+
+        @authorization_handler ||= authorization_handler_form_class.from_params(authorization_handler_params)
+      end
+
+      def signature_workflow_name
+        @signature_workflow_name ||= initiative&.type&.document_number_authorization_handler
+      end
+
+      def hash_id
+        return unless initiative && (unique_id || user)
+
+        @hash_id ||= Digest::MD5.hexdigest(
+          [
+            initiative.id,
+            unique_id || user.id,
+            Rails.application.secret_key_base
+          ].compact.join("-")
+        )
+      end
+
+      # The attributes of the handler that should be exposed as form input when
+      # rendering the handler in a form.
+      #
+      # Returns an Array of Strings.
+      def form_attributes
+        attributes.except("id", "user", "initiative").keys
+      end
+
+      # The String partial path so Rails can render the handler as a form. This
+      # is useful if you want to have a custom view to render the form instead of
+      # the default view.
+      #
+      # Example:
+      #
+      #   A handler named Decidim::CensusHandler would look for its partial in:
+      #   decidim/census/form
+      #
+      # Returns a String.
+      def to_partial_path
+        "decidim/initiatives/initiative_signatures/#{signature_workflow_name.sub(/_handler$/, "")}/form"
+      end
+
+      def self.requires_extra_attributes?
+        new.form_attributes.present?
+      end
+
+      private
+
+      # It is expected to validate that no other user has voted with the same
+      # unique_id and scope. The unique_id should be defined by the classes
+      # inherited from this taking a personal data attribute like a document
+      # number. If not defined the user id is used
+      def uniqueness
+        add_invalid_base_error if Decidim::InitiativesVote.exists?(scope:, hash_id:)
+      end
+
+      def valid_metadata
+        return if authorization_handler.blank?
+        return if authorization_handler.valid?
+
+        # Promote errors
+        if promote_authorization_validation_errors
+          keys = attributes.keys.map(&:to_sym)
+
+          authorization_handler.errors.each do |error|
+            next unless keys.include?(error.attribute.to_sym)
+
+            errors.add(error.attribute, error.type) unless document_number.to_s.end_with?("X")
+          end
+        end
+
+        add_invalid_base_error
+      end
+
+      def valid_authorized_scopes
+        return if authorized_scopes.present?
+
+        add_invalid_base_error
+      end
+
+      def add_invalid_base_error
+        errors.delete(:base)
+        errors.add(:base, I18n.t("invalid_data", scope: "decidim.initiatives.initiative_signatures.fill_personal_data"))
+      end
+
+      def encryptor
+        @encryptor ||= DataEncryptor.new(secret: encryption_secret)
+      end
+
+      def encryption_secret
+        "personal user metadata"
```

Comment:
Shouldn't we use an ENV variable for this?

Reply:
You are right, i copied it from the old handler but better to move the definition to a secret.